Migration (topic-editor-state.service.ts) AngularJS to Angular 8

Deepam101 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #8472 .
2. This PR does the following: Migrates topic-editor-state-service.ts from AngularJS to Angular 8

Video : https://user-images.githubusercontent.com/50206813/104214697-cd6b8700-545d-11eb-9b4e-dc6b25a857f3.mp4
## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Commit:
Initial migrations

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Initial changes

Commit:
Initial changes

Commit:
Fixed tests

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
changes

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'origin' into TopicEditorState

Commit:
changes

Commit:
Changes to saveTopic function

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
changes

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

srijanreddy98 commented on proto/text_classifier_pb2.py:
```diff
@@ -0,0 +1,70 @@
+# -*- coding: utf-8 -*-
```
Are these expected?

Deepam101 replied:
No they weren't expected. Might have popped from some earlier buggy pull

srijanreddy98 commented on core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts:
```diff
@@ -18,528 +18,544 @@
+    others: []
+  };
+  private _classroomUrlFragment = 'staging';
+  private _storySummariesInitializedEventEmitter = new EventEmitter();
```
Do this for the other three event emitters below.
```suggestion
  private _storySummariesInitializedEventEmitter: EventEmitter<void> = new EventEmitter();
```

Deepam101 replied:
Done

srijanreddy98 commented on core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts:
```diff
@@ -18,528 +18,544 @@
-      get onTopicInitialized() {
-        return _topicInitializedEventEmitter;
-      },
+  get onTopicInitialized(): EventEmitter<unknown> {
```
Do similar changes in this file as well.
```suggestion
  get onTopicInitialized(): EventEmitter<void> {
```

Deepam101 replied:
Done

srijanreddy98 commented on core/templates/pages/topic-editor-page/services/topic-editor-state.service.spec.ts:
```diff
@@ -16,191 +16,124 @@
-      return $q(function(resolve, reject) {
-        if (!self.failure) {
-          resolve(self.newBackendTopicObject);
+    fetchTopic():Promise<unknown> {
```
I think the unknown type is misused here. Can you try and find the actual type for this. Looks like a TopicBackendDict. Do this similarly for other functions as well.

Deepam101 replied:
Done

srijanreddy98 reviewed (changes requested):
@Deepam101 took a pass. PTAL!

Commit:
changes

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
changes

iamprayush commented:
Hey @Deepam101 any updates on this? Please make sure that you update the assignee list with the people you want a review from once you've addressed the comments. Thanks!

Deepam101 commented:
> Hey @Deepam101 any updates on this? Please make sure that you update the assignee list with the people you want a review from once you've addressed the comments. Thanks!

Yeah doing it right away!

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

srijanreddy98 reviewed (changes requested):
LGTM, only one concern, don't use `angular.copy` or anything apart from `angular.module`

sajalasati commented:
Hi @Deepam101! Can you please resolve the merge conflict? After addressing all comments you can leave a PTAL comment for the reviewer? You can follow the instructions for doing so in our [wiki](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#address-review-comments-until-all-reviewers-give-lgtm-looks-good-to-me). Thanks!

Deepam101 commented:
> Hi @Deepam101! Can you please resolve the merge conflict? After addressing all comments you can leave a PTAL comment for the reviewer? You can follow the instructions for doing so in our [wiki](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#address-review-comments-until-all-reviewers-give-lgtm-looks-good-to-me). Thanks!

Yes working on it!

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
changes

Commit:
changes

Deepam101 commented:
> LGTM, only one concern, don't use `angular.copy` or anything apart from `angular.module`

Done!

kevintab95 commented on core/templates/pages/topic-editor-page/navbar/topic-editor-navbar.directive.ts:
```diff
@@ -200,10 +200,11 @@ angular.module('oppia').directive('topicEditorNavbar', [
+              TopicEditorStateService.saveTopic(commitMessage).then(
+                () => {
+                  AlertsService.addSuccessMessage(
+                    'Changes saved.', 1000);
```
Do we need a `$rootScope.$applyAsync();` here?

Deepam101 replied:
Done!

kevintab95 reviewed (changes requested):
Thanks @Deepam101! One comment PTAL. Can you also do the following please?
1. Update your branch with develop (see [thread](https://groups.google.com/g/oppia-dev/c/k6ppgwvq3hI)).
2. Rerun the relevant e2e suites 3 consecutive times and post screenshots of the passing result on the PR. This is to make sure that we do not introduce functionality that can cause e2e flakes. You may need to request one of the reviewers for rerunning your tests if you don't have permission to rerun them yourself.

srijanreddy98 reviewed (approved):
The changes look good to me. Just one last comment. Find out all the functions in the service that return a promise. Then go to the places where these functions are used and add `$rootScope.$applyAsync()` at the end of the .then block.

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
Merge branch 'TopicEditorState' of https://github.com/Deepam101/oppia into TopicEditorState

Deepam101 commented:
> The changes look good to me. Just one last comment. Find out all the functions in the service that return a promise. Then go to the places where these functions are used and add `$rootScope.$applyAsync()` at the end of the .then block.

Done!

DubeySandeep commented:
@Deepam101, Is this PR ready for review? if yes, can you please ask reviewers to take a pass as suggested in the "PR pointers" section above?

> Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"

DubeySandeep commented:
@kevintab95 Can you please review this PR?

kevintab95 commented on core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts:
```diff
@@ -18,541 +18,556 @@
+   * async backend call was successful and that _topicWithUrlFragmentExists
+   * has been successfully updated.
+   */
+  updateExistenceOfTopicUrlFragment(
```
This does an async operation and therefore we had used a `$rootScope.$applyAsync();`. but that has been removed. How are we handling this now?

Deepam101 replied:
Made the function to return a promise. Every angularJS file where this function was called, I have added an `$rootScope.$applyAsync();`. This removes the requirement of using `$rootScope.$applyAsync();` in the function itself

kevintab95 commented on core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts:
```diff
@@ -18,541 +18,556 @@
+   * async backend call was successful and that _topicWithNameExists
+   * has been successfully updated.
+   */
+  updateExistenceOfTopicName(
```
This does an async operation and therefore we had used a `$rootScope.$applyAsync();`. but that has been removed. How are we handling this now?

Deepam101 replied:
Made the function to return a promise. Every angularJS file where this function was called, I have added an $rootScope.$applyAsync();. This removes the requirement of using $rootScope.$applyAsync(); in the function itself

kevintab95 commented on core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts:
```diff
@@ -18,541 +18,556 @@
+   * will clear the UndoRedoService of pending changes. This function also
+   * shares behavior with setTopic(), when it succeeds.
+   */
+  saveTopic(
```
This does an async operation and therefore we had used a `$rootScope.$applyAsync();`. but that has been removed. How are we handling this now?

Deepam101 replied:
Made the function to return a promise. Every angularJS file where this function was called, I have added an $rootScope.$applyAsync();. This removes the requirement of using $rootScope.$applyAsync(); in the function itself

kevintab95 commented on core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts:
```diff
@@ -18,541 +18,556 @@
+   * Loads, or reloads, the subtopic page stored by this service given a
+   * specified topic ID and subtopic ID.
+   */
+  loadSubtopicPage(topicId: string, subtopicId: number): void {
```
This does an async operation and therefore we had used a `$rootScope.$applyAsync();`. but that has been removed. How are we handling this now?

Deepam101 replied:
Made the function to return a promise. Every angularJS file where this function was called, I have added an $rootScope.$applyAsync();. This removes the requirement of using $rootScope.$applyAsync(); in the function itself

kevintab95 commented on core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts:
```diff
@@ -18,541 +18,556 @@
+   * specified topic ID. See setTopic() for more information on
+   * additional behavior of this function.
+   */
+  loadTopic(topicId: string): void {
```
This does an async operation and therefore we had used a `$rootScope.$applyAsync();`. but that has been removed. How are we handling this now?

Deepam101 replied:
Made the function to return a promise. Every angularJS file where this function was called, I have added an $rootScope.$applyAsync();. This removes the requirement of using $rootScope.$applyAsync(); in the function itself

kevintab95 reviewed (changes requested):
Thanks @Deepam101! Left few comments. PTAL!

sajalasati commented:
Hi @Deepam101, do you have any updates?

Deepam101 commented:
> Hi @Deepam101, do you have any updates?

Yes working on it!

Commit:
async changes

darksun27 commented on core/templates/domain/topic/editable-topic-backend-api.service.ts:
```diff
@@ -203,7 +203,7 @@ export class EditableTopicBackendApiService {
   private _updateTopic(
       topicId: string,
-      topicVersion: string,
+      topicVersion: number,
```
Are you sure about this change, because I am able to see strings are passed in version where the function is used.

Deepam101 replied:
https://github.com/oppia/oppia/blob/36dd0bd3121e59861a4e2a7f38f1a7f937f84a7a/core/templates/domain/topic/TopicObjectFactory.ts#L47

darksun27 commented on core/templates/domain/topic/editable-topic-backend-api.service.ts:
```diff
@@ -302,7 +302,7 @@ export class EditableTopicBackendApiService {
   updateTopic(
       topicId: string,
-      topicVersion: string,
+      topicVersion: number,
```
ditto.

Deepam101 replied:
https://github.com/oppia/oppia/blob/36dd0bd3121e59861a4e2a7f38f1a7f937f84a7a/core/templates/domain/topic/TopicObjectFactory.ts#L47

darksun27 commented on core/templates/pages/topic-editor-page/editor-tab/topic-editor-tab.directive.spec.ts:
```diff
@@ -249,10 +249,16 @@ describe('Topic editor tab directive', function() {
   });
 
-  it('should call the TopicUpdateService if name is updated', function() {
+  fit('should call the TopicUpdateService if name is updated', function() {
```
did you force push this PR?

Deepam101 replied:
oh yeah it needed some changes..ill be converting it to a draft for some time

darksun27 commented on core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts:
```diff
@@ -18,541 +18,562 @@
-        return angular.copy(_groupedSkillSummaries);
-      },
+  getGroupedSkillSummaries(): unknown {
+    return cloneDeep(this._groupedSkillSummaries);
```
clone deep is used to created a separate object, I don't think so the return type would be unknown here.

Deepam101 replied:
yes done

darksun27 commented on core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts:
```diff
@@ -18,541 +18,562 @@
-      setTopic: function(topic) {
-        _setTopic(topic);
-      },
+  getCanonicalStorySummaries(): unknown[] {
```
should unknown here be StorySummary[] ?

Deepam101 replied:
yes done

darksun27 reviewed (commented):
Thanks, @Deepam101. Took a pass some comments.

Commit:
mentioned changes

Commit:
changes

Commit:
type change in testing

darksun27 reviewed (approved):
LGTM! Thanks, @Deepam101!

Deepam101 commented:
> Assigning @kevintab95 for code owner reviews. Thanks!
@kevintab95 PTAL!

kevintab95 reviewed (approved):
LGTM, thanks @Deepam101!

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

DubeySandeep commented:
@Deepam101 Can you please update this PR with upstream/develop to fix the lighthouse issue?

Deepam101 commented:
> @Deepam101 Can you please update this PR with upstream/develop to fix the lighthouse issue?

yeah on it!

Commit:
Merge remote-tracking branch 'upstream/develop' into TopicEditorState

Commit:
changes

Commit:
changes
